rep-0144.rst

```diff
@@ -0,0 +1,129 @@
+REP: 144
+Title: ROS Package Naming and Versioning
+Author: Vincent Rabaud <vincent.rabaud@gmail.com>
+Status: Draft
+Type: Informational
+Content-Type: text/x-rst
+Created: 28-Jan-2015
+Post-History:
+
+
+Abstract
+========
+
+This REP gives advice on how a ROS package should be named and versioned. It
+formalizes and extends conventions that were formerly described in [1]_.
+
+Motivation
+==========
+
+As the number of ROS packages increases, it is hard to quickly find a package
+and guess its functionality based on its name. Over time, the lack of naming 
+conventions created problems like use of unexplained acronyms or
+packages with the same functionality but different names.
+
+This REP proposes rules to name a ROS package properly. Those 
+rules are not mandatory, merely informative.
+
+
+Package Naming
+==============
+
+Global Rules
+------------
+
+* ``alphanumerics`` are ``A-Za-z0-9`` only
+* a package name is defined more strictly than with Debian at [2]_. It must only consist
+  of lower case letters ``a-z``, digits ``0-9`` and ``_`` separators. They must be at
+  least two characters long and must start with an alphanumeric character
+* package names should be specific enough to identify what the package does. For example,
+  a motion planner should not called ``planner``. If it implements the wavefront propagation
+  algorithm, it might be called ``wavefront_planner``. There's obviously tension between
+  making a name specific and keeping it from becoming overly verbose
+* using catchall names such as ``utils`` should be avoided as they do not scope what goes
+  into the package or what should be outside the package
+* a package name should not contain ``ros`` as it is redundant. Exceptions include 
+  core packages and ROS bindings of an upstream library (e.g. ``moveit_ros``)
+* one of ROS's goals is to develop a canonical set of tools for making robots do
+  interesting things. The package name should describe what the package does, not where
+  it came from
```

Comment:
tl;dr I would propose that the recommend be to pick a descriptive name and then prefix it with the "where it came from" until such point that it is truly commonly used and owned.

Maybe it is sufficient to say that the rule below, "**if a package is specialized by an entity (lab, company, ...), prepend the name of the entity**", should be considered true until something is commonly used, owned, and maintained.

----

This is something that I have always had trouble with. I understand the intention of wanting to have packages which invoke common ownership, i.e. `navigation` versus `wg_navigation`. However, I think that it can also be misleading. Consider the case where some grad student can create a package called `wavefront_planner` for a pet project, after which he releases it, and then graduates never to touch it again. At this point we have something which is very specifically named. To me, this high visibility name then implies to uninformed users that the community has reached some sort of consensus and that this package is ***the*** `wavefront_planner` package to use, but in fact it is a stale, highly specialized package which is not getting updated. This leads to all kinds of silly things like `fast_light_wavefront_planner` and `planner_wavefront` , all names which come from people trying to innovate in the same crowded namespace. In this case I think something like `<github_name>_wavefront_planner` or `<university_name>_wavefront_planner` would be much more appropriate, because it is less polluting to the namespace, it is still easily searched for, and it is (more) self descriptive. When and if a package becomes commonly enough used, there will be sufficient motivation that the name could be changed to something more official.

From my perspective, a package with a single author (person or organization) should describe where it came from until which point collaboration and maturity increase to a point to where it is *the* implementation that everyone uses and the community depends on, e.g. `amcl`.

The justification for this point from the wiki is:

> An Ubuntu distribution offers approximately 33,000 packages without inserting origin or authorship into names.

I think this is flawed reasoning, first because the packages in Ubuntu are packaged not by the people who wrote the software (in most cases) but by Ubuntu/Debian packager's. Second there are name collisions all the time and names are regularly modified (from upstream) to make them fit, e.g. the `python-` prefix, and docker is actually `docker.io`, etc... Finally, we have many more one-off packages and experimental packages, due to the ad hoc style of development that goes on in robotics, which is something that Ubuntu does not have to deal with because they use PPA's which commonly have name collisions and simply cannot be used together at the same time.